### DISCUSSION TITLE
How to make an automation script for running the OpenRadioss?

### DISCUSSION THREAD

**Discussion by Majid-MHZ** [OpenRadioss/OpenRadioss#3178] [Contributors]
Hello everyone,

I have a question related to my project (powder compaction forming) which is going to be done using OpenRadioss.
I simulate the process in LS-PrePost and use OpenRadioss to run the extracted .k file. The entire process is divided into two steps: 1) loading (compaction) and 2) unloading (removing the applied load on the punch). A velocity boundary condition is defined to move the punch and apply the load.
I would like to compact the particles (by moving the punch downward) until I reach a specific pressure during the compaction step. Once the pressure reaches the predetermined level, I want to remove the applied load and move the punch upward during the unloading step.
I would like to know how I can manage this process effectively. Is there a way to script OpenRadioss to calculate the reaction force during the simulation, or is there another method that can handle this process? 
Any suggestions or recommendations would be greatly appreciated. Please let me know if you have any questions.

Thank you for your kind help in advance.

Best regards,
Majid

**Comment by SyazwanSamad** [comment]
It seems to be complicated to fully define the model with the LS PrePost environment.

Let me tell you my experience. I have the model prepared using LS-Prepost. I developed a python script to automatically create OpenRadioss include file and automatically append it into the main LS Dyna file (using *Include_radioss keyword). The reason I used this approach is because the LS Dyna keywords that I wanted to use is not supported by OpenRadioss.

In your case, I would also do the similar thing; After generating the model using LS PrePost, then I'll define /GAUGE to measure the pressure. Then I'll define /SENSOR/GAUGE that can control the start and stop time of the two /IMPVEL applied on the punch (the first /IMPVEL is to apply velocity towards the powder and second one is for the unloading velocity).

**Reply by Majid-MHZ** [reply]
Dear @SyazwanSamad, thank you for your reply. Could you please explain the procedure in a simple word? Unfortunately, I couldn't get you well. Besides the above-mentioned issue, I have faced another problem. Do you have any experience regarding the following problem?

I have already completed the loading process without any issues, following the procedure mentioned above. Now, I would like to simulate a loading-unloading process, but I have encountered a problem with defining the step time for the unloading stage. To the best of my knowledge, it is impossible to define two steps with different conditions, such as time step, mass scaling, and so on using Ls-PrePost, as can be done in ABAQUS. What I found is that I should use the *RESTART option to perform the unloading stage. I have searched the Internet, but I haven’t been able to figure it out yet. My question is: how should I use this option in my simulation? Based on your experience, what would be the best solution?

**Reply by SyazwanSamad** [reply]
It is not really easy to explain the process if you don't know OpenRadioss keywords well. I have given the necessary keywords; I suggest you check the manual to understand further.
1. /IMPVEL - it is basically the same as *BOUNDARY_PRESCRIBED_MOTION with velocity option. The OpenRadioss version has the advantage where it can be controlled by sensors. https://2021.help.altair.com/2021.1/hwsolvers/rad/topics/solvers/rad/impvel_starter_r.htm
2. /GAUGE - I believe there's no equivalent keyword in LS Dyna because /GAUGE output is pressure, not forces. https://2021.help.altair.com/2021/hwsolvers/rad/topics/solvers/rad/gauge_starter_r.htm
3. /SENSOR/GAUGE - A sensor that is activated /deactivated based on pressure. https://2021.help.altair.com/2021/hwsolvers/rad/topics/solvers/rad/sensor_starter_r.htm

As for your 2nd question, both LS Dyna and OpenRadioss can perform restart operation.
In LS Dyna, you can use sw1 switch to stop and generate the d3dump file. Then create a new input file e.g restart.k that contain the changes of the time step/mass scaling, output interval, (even delete contacts and parts). To continue just run 'lsdyna i=restart.k r=d3dump01'; assuming this is the 1st dump.

In OpenRadioss, control file can be used to generate a restart file. https://2021.help.altair.com/2021/hwsolvers/rad/topics/solvers/rad/control_file_c_file_r_2.htm
Then create a second engine file (*_0002.rad) which contain the changes of the timestep, output interval etc. from there just run the OpenRadioss (engine) with the newly created engine file.

**Comment by PaulAltair** [comment]
This is a good candidate to use the new /FUNCT_PYTHON feature I think: It is already implemented in OpenRadioss 

Using this, you should be able to define your loading function with reference to your pressure directly in a single run

It isn't in the online documentation yet (will be in the commercial 2025 release): for now, see: 
[FUNCT_PYTHON.pdf](https://github.com/user-attachments/files/18048101/FUNCT_PYTHON.pdf)

Code for pressure: 
![image](https://github.com/user-attachments/assets/b0255780-807f-4145-b2e8-5420be8b9c9b)


This model is an example of how you might use it

[compression.zip](https://github.com/user-attachments/files/18048104/compression.zip)

When running starter, you need to append '-python' to the starter run command, or, if you are using the openradioss gui, append it to the starter command line in runopenradioss.py: 
`starter_command = [os.path.join(self.openradioss_path, starter_exec), "-i", self.file, "-nt", self.nt, "-np", self.np, "-python"] 
`

If you can share a small version of your input? then we may help you get the correct syntax?

**Reply by Majid-MHZ** [reply]
Dear @PaulAltair,
Thank you for your kind reply.

I will prepare it and send it to you as soon as possible.
I appreciate your help once again.

Best regard,
Majid

**Comment by Majid-MHZ** [comment]
Dear @PaulAltair,
I hope you are doing well.

I have uploaded a sample file of my simulation to this message. Please check it and let me know if you need further information.
I use a single .k file for the simulation. When the time reaches the checkpoint, the second stage (unloading) of the simulation starts (Command file). 
[Input_file.zip](https://github.com/user-attachments/files/18072450/Input_file.zip)

**Comment by PaulAltair** [comment]
Sorry for delay, I only just saw your reply, I will take a look this week

**Comment by Majid-MHZ** [comment]
Dear @PaulAltair,

Thank you for your message.
I am looking forward to hearing from you. 

Best regard,
Majid

**Reply by PaulAltair** [reply]
Ok, I've had a quick look, your timestep of 1e-4 was much too large (the mass was being scaled by a factor of 60 million). Also as sent, you had an initial gap between punch and particles of over 0.1mm (and a gap between particles and die) with the applied velocity, the punch will only move 0.1mm so not compress anything. I have moved the punch closer to the particles and the particles closer to the die, and reduced timestep from 1e-4 to 1e-7 to check the initial phase. The modified .k is attached below the image. What is your aim after this? to Reverse the load at a given force?

![image](https://github.com/user-attachments/assets/78629e45-eb2e-42fd-bdab-ac83141ca8dc)
[Sample.zip](https://github.com/user-attachments/files/18163919/Sample.zip)

**Reply by Majid-MHZ** [reply]
Dear @PaulAltair,
I hope you are doing well.

Thank you for your help and support. I agree with you regarding the time step, but the simulation takes a long time once I set it to 1e-7. As the number of particles is around 1000 at most in the actual simulation, I set the time step to 1e-4. I searched for ways to speed up the simulation, but I couldn't find anything except the parameters mentioned above. Do you have any suggestions on how I can speed up the simulation while considering small time steps like 1e-7?
Additionally, this simulation aims to perform a loading-unloading process—namely, applying a load up to a specific pressure, reversing the load once it reaches the predetermined pressure, and moving the punch upwards until the pressure comes down to zero. Finally, I need to calculate the relative density of the compacted particles.

Please let me know if you need further information.

Best regards,
Majid

**Reply by PaulAltair** [reply]
The size of the elements and stiffness of the material determine the stable timestep based on courant principle

smallest element size / (sqrt (E / rho))

so, there is not much you can do to speed it up other than using more CPU (I ran it on 96 cpu and it took 55 minutes), using larger elements or, accepting more added mass (even at 9e-8 the added mass is 60 x though, so I wouldn't go any bigger probably)

**Reply by PaulAltair** [reply]
ok, I experimented, it seems, the added mass is not so critical, I tried increasing and the results do not change very much

So with timestep of 1.5e-6 I was able to run in just over 1hr on my laptop (8 processes) for the pressure = 250MPa case here

[PS_Sample_250.zip](https://github.com/user-attachments/files/18176904/PS_Sample_250.zip)

This is the setup:

in Sample.k, there is added a *INCLUDE_RADIOSS card calling the file Extra_Radioss_Cards.rad

![image](https://github.com/user-attachments/assets/723aa1a5-9d7b-459d-9f40-c195d22f63af)

Inside it are defined 2x Sensors for the Rigid Body force in the Die (Punch and Die forces match exactly): 

SENSOR 1 is set to be activated at a max force ( Analagous to Pressure, since area is 1x1 in this case) of 250MPa (this is where you may change that value for max pressure)

SENSOR 2 is set to be activated at a min force of 0.1MPa (to detect when the load has been removed)

![image](https://github.com/user-attachments/assets/a2948edd-68c7-4805-95de-7df3327dc710)

Also inside Sample.k I set the termination time to 0.0 (so the automatically created 0001.rad will run for just one cycle) and set the velocity for the punch to -0.2 constant

then in the attached 0002.rad and 0003.rad engine files: (here timestep is set 1.5e-6)

The 0002.rad stops the job when the first sensor is triggered (at 250MPa)

Then the 0003.rad restarts the run, modifying the loading curve to +0.08 velocity and resetting the SENSOR 2 (since it is triggered already at start of run before there is any contact) then the job will stop at the end, when SENSOR 2 is again triggered (Pressure < 0.1MPa)

![image](https://github.com/user-attachments/assets/e3700ad3-f5f1-480b-94bc-dc04f21a899b)

Below are results, e.g. for 150MPa ,250MPa and 1000MPa

![image](https://github.com/user-attachments/assets/defdd8e2-4887-44fa-bd7a-138c3bd2e0f9)

I hope this helps you, You may experiment here to use different timesteps and pressures

Good Luck,

Paul

**Reply by Majid-MHZ** [reply]
Dear @PaulAltair ,
I hope you are doing well.

Thank you for your thorough explanation of the simulation procedure; I truly appreciate it. I have a question that I'd like to ask.
As you mentioned, the loading process stops at 250 MPa when the area of the punch or die is 1 x 1 mm, which is acceptable in this case. However, how should I manage the situation if the area is different, for example, 0.5 x 0.5 mm? If the punch area is 0.5 x 0.5 mm and I aim to conduct the compaction process up to 400 MPa, what adjustments should I consider?

Best regards,
Majid

**Reply by PaulAltair** [reply]
the sensor is monitoring reaction force on the rigid body, so, in case the area was 0.5mm by 0.5mm (0.25mm^2) the total force would be 1/4 the Pressure in effect (i.e. for 400MPa, the force to stop the job would only be 100N in this case, 100N/0.25mm^2 = 400MPa)

**Reply by Majid-MHZ** [reply]
I got how it works now. Thank you for your valuable guidance. Could you please explain the contactdifinitions that you have defined in the sample.k file? I believe you have noticed that I defined the general contact incorrectly. I would like to define a general contact specifically between the particles (excluding the die and punch from this contact). Could you please guide me on how to configure it so that it only considers contact between particles? Also, I need to define a contact between the punch-particles and die-particles. To do that I seleceted all particles and defined a surface as All particles. Then, I defined a surface-to-surface contact between the rigid bodies and the created surface. I want to know if what I have done is correct or not.
![image](https://github.com/user-attachments/assets/572901a0-bbd1-4701-835e-3e25ece4d834)

Best wishes,
Majid

**Comment by PaulAltair** [comment]
Yes, I reversed your main-secondary pairs (main coarser as rule of thumb) as you had penetrations happening in your original model. There is a bit of a mixture of settings left there, partially based on your original file, but what there is currently in the above

For the Contact Automatic General, that is your particle to particle self contact, it uses SSTYP 6 (exclude part list) and Part Set 301 is the Die AND Punch, so, that is a self contact of particles only (you could also use SSID 303 and SSTYP 2, that would explicitly include the particles instead of 'everything except the die and punch' the advantage to doing it as an exclude set like it is now, is that any new particles you added would automatically be included in the contact)

The first Automatic Surface is The Die and Punch part set to the particle part set, and the 2nd Automatic Surface to Surface is a Segment set of the punch to the Particle Part Set

I've made a revised version here: So now the General Contact is unchanged, but the 2 Surf - Surf Contacts are Part based Die - Particles and Punch - Particles (see above, you could use SSID 301 and SSTYP 6 for both of these 2 if you wanted, it would make the setup more general)

![image](https://github.com/user-attachments/assets/3afb406c-ceed-40a3-a50c-81648c863b32)

[PS_TEST_UPDATE_CONTACTS.zip](https://github.com/user-attachments/files/18201734/PS_TEST_UPDATE_CONTACTS.zip)

**Reply by PaulAltair** [reply]
NB: This one also has pressure set at 1000 :)

**Reply by Majid-MHZ** [reply]
Dear @PaulAltair,
I hope you are doing well.

Based on our previous discussion, and as you may remember, I am working on powder compaction simulation using OpenRadioss. I plan to run the simulation at different pressures. To achieve this, you suggested using a sensor method so that when the force meets the predetermined pressure, the loading simulation is switched to an unloading process. This proposed method is excellent and works perfectly at low pressures, but I encounter an unusual result when I increase the pressure. I conducted a simulation at 1400 MPa, keeping the other conditions the same with those from previous low-pressure simulations. I set the time step to 1.5E-6, as used for the low pressure. However, I faced two problems. First, the second sensor did not terminate the simulation once the pressure reached around 0 MPa during the unloading step. The second issue is that the particles in contact with the punch seem to be glued to the punch surface and move upward with it. I can provide a captured movie of the obtained result if you provide a transfer link. I couldn't upload here due to the file size issue.

Do you know why I get such a result? Could you please assist me on how I should fix these problems?

Thank you for your support in advance.

Best regards,
Majid
<img width="874" alt="11" src="https://github.com/user-attachments/assets/1aebbff8-5198-45fe-90c1-c8a233ade935" />
<img width="617" alt="12" src="https://github.com/user-attachments/assets/a90fe3ca-1288-4e14-b4a6-2c06b35c0f5a" />
<img width="640" alt="13" src="https://github.com/user-attachments/assets/20d19dd2-4442-4557-a170-c9c25aaa0eb3" />
<img width="964" alt="14" src="https://github.com/user-attachments/assets/4889d825-79eb-4fa9-8fe1-13fc69166b12" />

**Comment by Majid-MHZ** [comment]


**Reply by PaulAltair** [reply]
I have stress/strain in the anim files: but confirm that after anim-d3plot the stresses are not there in the d3plot, I guess it is something in the anim-d3plot tool, I will try to investigate
![image](https://github.com/user-attachments/assets/da6c1702-a598-4207-9215-19a4fb96d3c4)

**Reply by PaulAltair** [reply]
in fact the anim-d3plot tool is not converting solid stresses/strains yet, so this is expected, but you wouldn't have been getting them with your original run either?

**Reply by PaulAltair** [reply]
This is quite odd, in fact, there is no mechanism today so far as I am aware, to generate solid stresses in the d3plot, they are not written, so, I do not know how you have them for any run from OpenRadioss it ought not be possible, what you show here is not what you shared originally though, you don't have coating shells? And this was run OpenRadioss with d3plot from anim2d3plot? I am not sure what you mean by DATASET?

**Reply by SyazwanSamad** [reply]
@Majid-MHZ 
The original anim to d3plot does not convert solid stresses to d3plot.Please check if you are using the one that I modified; which can convert the stresses.

**Reply by PaulAltair** [reply]
Hi @SyazwanSamad did you push your updates back to vortexradioss?

**Reply by SyazwanSamad** [reply]
@PaulAltair 
No, I did not push it to Vortex-radioss because the modification is made for my specific application (example DAMA is mapped to history variable to extra history variable #1, VONM to variable #2. In this case DAMA need to be defined in order to map the VONM. OR if TENS/STRESS are available in ANIM, it will be mapped as stresses in d3plot). 

I remember that I shared the modified code somewhere in one of the discussion, @Majid-MHZ could be picking my code from the discussion. For your information, this is the location where I stored the modified Vortex-radioss.
https://github.com/SyazwanSamad/Vortex-Radioss
